Add reply-comments.test, extend comments.model, extend comments.hooks

Mastercuber opened the pull request:
How can I access the "children" in the Comments.vue file or how to attach the children to one comment? The schema should get the appropriate child comments.

Commit:
Add reply-comments.test, extend comments.model, extend comments.hooks, open commentForm on reply

roschaefer commented on test/services/reply-comments.test.js:
```diff
@@ -0,0 +1,91 @@
+      comment = await service.create(commentData, params);
+
+      replyData.parentCommentId = comment._id;
+      const reply = await service.create(replyData, params);
```
You create a reply by the original author here. `params` does not have the id of the replyAuthor. Add a new test case and `assert.equal(reply.userId, replyUser._id)`.

Mastercuber replied:
Got it! I will change it

roschaefer commented on server/models/comments.model.js:
```diff
@@ -7,6 +7,7 @@ module.exports = function (app) {
     userId: { type: String, required: true, index: true },
     contributionId: { type: String, required: true, index: true },
+    parentCommentId: { type: String, default: '' },
     content: { type: String, required: true },
```
We decided to have only one level of nesting: Contribution >> Comment >> Reply. Instead of a self reference, we could also create a new model `replies`. What is your opinion on it, @appinteractive ?

Mastercuber replied:
done with `parentcommentId` and a comment.hook for getting the children per comment

roschaefer commented on server/models/comments.model.js:
```diff
@@ -7,6 +7,7 @@ module.exports = function (app) {
   const comments = new mongooseClient.Schema({
     userId: { type: String, required: true, index: true },
     contributionId: { type: String, required: true, index: true },
+    parentCommentId: { type: String, default: '' },
```
Why `default: ''` empty string?

Mastercuber replied:
I removed the default

roschaefer reviewed (changes requested):
@Mastercuber great work, you can reference the issue you want to close in the description. Too bad, you cannot use Github's autoclose feature: https://help.github.com/articles/closing-issues-using-keywords/

I don't know myself how to automagically include associations in a feathersjs response. A quick research showed there is support for sequelize but I don't know if there is support for mongodb. If you do the self-reference (as you did in the PR) I think you have to implement it yourself with feathers' hooks: https://docs.feathersjs.com/api/hooks.html

Commit:
add another test to reply-comments.test, extend comments.hooks commentSchema, add nested comments with standard depth 1

Commit:
add right color to triangle left of comment; insert Mention as real Mention; fix scrolling

Commit:
fix eslint error

roschaefer commented on test/services/reply-comments.test.js:
```diff
@@ -0,0 +1,102 @@
+
+    contribution = await contributionService.create(contributionData, params);
+
+    commentData.userId = user._id;
```
Setting the `userId` on a comment or a reply should not be necessary. The user authenticates with the server, so the server should be responsible for the `userId` field.

Mastercuber replied:
I could change this in the `comments.test.js`, but without it the `reply-comments.test.js` will fail when running it.

roschaefer commented on test/services/reply-comments.test.js:
```diff
@@ -0,0 +1,102 @@
+    delete commentData.userId;
+    delete commentData.contributionId;
+    delete replyData.userId;
+    delete replyData.contributionId;
```
I think you don't need to reset the shared data here. Probably you are afraid of some tests interfering with other tests because of stale data? There are different approaches here, but I will share mine:

I make use of nested scenarios. If more than one test needs some data, I write `context('given a reply user', () => {`, put the `let` statement within the block and setup the user in a `beforeEach` block. I don't reset the data. You will quickly forget to reset all your data anyways. But within the context block, all tests should be OK with the preexisting data. If some tests do not require the data to be setup, they don't belong into that block.

Do you see what I mean?

Mastercuber replied:
Ok I can see the point. It's also good for the test performance if the data is not created per test, but per context.

Mastercuber commented on test/services/reply-comments.test.js:
```diff
@@ -0,0 +1,102 @@
+    user = await userService.create(userData);
+    replyUser = await userService.create(replyUserData);
+
+    params = {
```
yes, i changed it

Commit:
fixes to review comments

roschaefer committed:
Add test cases for `get` and `find` + refactoring

Commit:
Finish test

Commit:
Replace `assert.equal`

while it is deprecated [Info](https://nodejs.org/api/assert.html#assert_assert_equal_actual_expected_message)

Commit:
replace comments in for loop with computed value (double comment fix + linter fix)

roschaefer committed:
Merge remote-tracking branch 'origin/develop' into develop

Commit:
merge develop into fork

Commit:
Merge remote-tracking branch 'origin/develop' into develop

Commit:
fix while changing to strictEquals
